Nightly version doesn't work on Windows

PyCharm detects `COM3` as a device path for Raspberry Pi Pico, trying to open the REPL raises ...

```
java.util.concurrent.ExecutionException: Failed to start [C:\Users\Pavel.Karateev\.virtualenvs\test_micropython_2\Scripts\python.exe, C:\Users\Pavel.Karateev\git\intellij-micropython\build\idea-sandbox\plugins\intellij-micropython/scripts/microrepl.py, COM3] in COM3 [no such directory], [null], envs={}
	at org.jetbrains.plugins.terminal.LocalTerminalDirectRunner.createProcess(LocalTerminalDirectRunner.java:264)
	at com.jetbrains.micropython.repl.MicroPythonReplManager.startREPL(MicroPythonReplManager.kt:69)
	at com.jetbrains.micropython.repl.ToolWindowReplTab$replStartAction$1.actionPerformed(ToolWindowReplTab.kt:94)
	at com.intellij.openapi.actionSystem.impl.ActionButton.actionPerformed(ActionButton.java:227)
	at com.intellij.openapi.actionSystem.impl.ActionButton.lambda$performAction$2(ActionButton.java:206)
	at com.intellij.openapi.actionSystem.ex.ActionUtil.performDumbAwareWithCallbacks(ActionUtil.java:362)
	at com.intellij.openapi.actionSystem.impl.ActionButton.performAction(ActionButton.java:206)
	at com.intellij.openapi.actionSystem.impl.ActionButton.processMouseEvent(ActionButton.java:520)
	at java.desktop/java.awt.Component.processEvent(Component.java:6422)
	at java.desktop/java.awt.Container.processEvent(Container.java:2266)
	at java.desktop/java.awt.Component.dispatchEventImpl(Component.java:5027)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2324)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4855)
	at java.desktop/java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4954)
	at java.desktop/java.awt.LightweightDispatcher.processMouseEvent(Container.java:4581)
	at java.desktop/java.awt.LightweightDispatcher.dispatchEvent(Container.java:4522)
	at java.desktop/java.awt.Container.dispatchEventImpl(Container.java:2310)
	at java.desktop/java.awt.Window.dispatchEventImpl(Window.java:2808)
	at java.desktop/java.awt.Component.dispatchEvent(Component.java:4855)
	at java.desktop/java.awt.EventQueue.dispatchEventImpl(EventQueue.java:794)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:739)
	at java.desktop/java.awt.EventQueue$3.run(EventQueue.java:733)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:97)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:766)
	at java.desktop/java.awt.EventQueue$4.run(EventQueue.java:764)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:399)
	at java.base/java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:86)
	at java.desktop/java.awt.EventQueue.dispatchEvent(EventQueue.java:763)
	at com.intellij.ide.IdeEventQueue.defaultDispatchEvent(IdeEventQueue.kt:690)
	at com.intellij.ide.IdeEventQueue.dispatchMouseEvent(IdeEventQueue.kt:638)
	at com.intellij.ide.IdeEventQueue._dispatchEvent(IdeEventQueue.kt:592)
	at com.intellij.ide.IdeEventQueue.access$_dispatchEvent(IdeEventQueue.kt:67)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:369)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1$1.compute(IdeEventQueue.kt:368)
	at com.intellij.openapi.progress.impl.CoreProgressManager.computePrioritized(CoreProgressManager.java:787)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:368)
	at com.intellij.ide.IdeEventQueue$dispatchEvent$processEventRunnable$1$1.invoke(IdeEventQueue.kt:363)
	at com.intellij.ide.IdeEventQueueKt.performActivity$lambda$1(IdeEventQueue.kt:997)
	at com.intellij.openapi.application.TransactionGuardImpl.performActivity(TransactionGuardImpl.java:113)
	at com.intellij.ide.IdeEventQueueKt.performActivity(IdeEventQueue.kt:997)
	at com.intellij.ide.IdeEventQueue.dispatchEvent$lambda$7(IdeEventQueue.kt:363)
	at com.intellij.openapi.application.impl.ApplicationImpl.runIntendedWriteActionOnCurrentThread(ApplicationImpl.java:861)
	at com.intellij.ide.IdeEventQueue.dispatchEvent(IdeEventQueue.kt:405)
	at java.desktop/java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:207)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:128)
	at java.desktop/java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:117)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:113)
	at java.desktop/java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:105)
	at java.desktop/java.awt.EventDispatchThread.run(EventDispatchThread.java:92)
Caused by: com.pty4j.windows.conpty.LastErrorExceptionEx: CreateProcessW failed: GetLastError() returned 267: The directory name is invalid.
	at com.pty4j.windows.conpty.ProcessUtils.start(ProcessUtils.java:91)
	at com.pty4j.windows.conpty.ProcessUtils.startProcess(ProcessUtils.java:25)
	at com.pty4j.windows.conpty.WinConPtyProcess.<init>(WinConPtyProcess.java:42)
	at com.pty4j.PtyProcessBuilder.start(PtyProcessBuilder.java:139)
	at org.jetbrains.plugins.terminal.LocalTerminalDirectRunner.createProcess(LocalTerminalDirectRunner.java:258)
	... 50 more
```

--- a/src/main/kotlin/com/jetbrains/micropython/actions/MicroPythonAction.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/actions/MicroPythonAction.kt
@@ -10,10 +10,9 @@ abstract class MicroPythonAction : AnAction() {
         val project = e.project ?: return
         val facet = project.firstMicroPythonFacet
         if (facet != null) {
-            e.presentation.isEnabled = facet.checkValid() == ValidationResult.OK
+            e.presentation.isEnabledAndVisible = facet.checkValid() == ValidationResult.OK
         } else {
-            e.presentation.isVisible = false
-            e.presentation.isEnabled = false
+            e.presentation.isEnabledAndVisible = false
         }
     }
 }

--- a/src/main/kotlin/com/jetbrains/micropython/actions/RunMicroReplAction.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/actions/RunMicroReplAction.kt
@@ -16,26 +16,27 @@
 
 package com.jetbrains.micropython.actions
 
+import com.intellij.openapi.actionSystem.ActionUpdateThread
 import com.intellij.openapi.actionSystem.AnActionEvent
-import com.intellij.openapi.actionSystem.CommonDataKeys
+import com.intellij.openapi.components.service
 import com.intellij.openapi.fileEditor.FileDocumentManager
 import com.intellij.openapi.fileEditor.ex.FileEditorManagerEx
 import com.intellij.openapi.module.Module
-import com.intellij.openapi.module.ModuleManager
 import com.intellij.openapi.module.ModuleUtil
 import com.intellij.openapi.vfs.VirtualFile
-import com.intellij.openapi.wm.ToolWindowManager
 import com.jetbrains.micropython.repl.MicroPythonReplManager
 import com.jetbrains.micropython.settings.firstMicroPythonFacet
 
 class RunMicroReplAction : MicroPythonAction() {
+
+  override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.EDT
 
   override fun actionPerformed(e: AnActionEvent) {
     val project = e.project ?: return
     val editor = FileEditorManagerEx.getInstanceEx(project).selectedTextEditor ?: return
 
     /*
-    Here we make best effort to find out module which is relevant to the current event.
+    Here we make our best to find out module which is relevant to the current event.
     There are two cases to consider:
     1) There is an open file present.
     2) No files are opened.
@@ -48,9 +49,6 @@ class RunMicroReplAction : MicroPythonAction() {
       project.firstMicroPythonFacet?.module
     }
 
-    if (module != null) {
-      MicroPythonReplManager.getInstance(module).startREPL()
-      ToolWindowManager.getInstance(project).getToolWindow("MicroPython")?.show()
-    }
+    project.service<MicroPythonReplManager>().startOrRestartRepl()
   }
 }

--- a/src/main/kotlin/com/jetbrains/micropython/repl/MicroPythonReplManager.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/repl/MicroPythonReplManager.kt
@@ -1,112 +1,24 @@
 package com.jetbrains.micropython.repl
 
 import com.intellij.openapi.components.Service
-import com.intellij.openapi.module.Module
-import com.intellij.openapi.module.ModuleUtil
-import com.jediterm.terminal.TtyConnector
-import com.jetbrains.micropython.settings.MicroPythonFacet
-import com.jetbrains.micropython.settings.microPythonFacet
-import com.jetbrains.python.sdk.baseDir
-import com.pty4j.PtyProcess
-import org.jetbrains.plugins.terminal.LocalTerminalDirectRunner
-import org.jetbrains.plugins.terminal.ShellStartupOptions
+import com.intellij.openapi.project.Project
+import com.intellij.util.messages.Topic
 
-interface CommsEventListener {
-    fun onProcessStarted(ttyConnector: TtyConnector)
-    fun onProcessDestroyed()
-    fun onProcessCreationFailed(reason: String)
+
+interface MicroPythonReplControl {
+    fun stopRepl()
+    fun startOrRestartRepl()
 }
 
-class MicroPythonReplManager(private val module: Module) {
-    private var listeners: MutableList<CommsEventListener> = mutableListOf()
-    private var currentProcess: Process? = null
-    private var currentConnector: TtyConnector? = null
+@Service(Service.Level.PROJECT)
+class MicroPythonReplManager(private val project: Project) : MicroPythonReplControl {
+    override fun stopRepl() =
+        project.messageBus.syncPublisher(MICROPYTHON_REPL_CONTROL).stopRepl()
 
-    companion object {
-        fun getInstance(module: Module): MicroPythonReplManager =
-                module.getService(MicroPythonReplManager::class.java)
-    }
 
-    fun startREPL() {
-        if (isRunning) {
-            stopREPL()
-        }
+    override fun startOrRestartRepl() =
+        project.messageBus.syncPublisher(MICROPYTHON_REPL_CONTROL).startOrRestartRepl()
 
-        val facet = module.microPythonFacet ?: return
-        val devicePath = facet.getOrDetectDevicePathSynchronously()
-
-        if (facet.pythonPath == null) {
-            notifyProcessCreationFailed("Valid Python interpreter is needed to start a REPL!")
-            return
-        }
-
-        if (devicePath == null) {
-            notifyProcessCreationFailed("Device path is not specified, please check settings.")
-            return
-        }
-
-        val initialShellCommand = mutableListOf(
-            facet.pythonPath!!,
-            "${MicroPythonFacet.scriptsPath}/microrepl.py",
-            devicePath
-        )
-
-        val terminalRunner = object : LocalTerminalDirectRunner(module.project) {
-            override fun getInitialCommand(envs: MutableMap<String, String>): MutableList<String> {
-                return initialShellCommand
-            }
-
-            fun getTtyConnector(process: PtyProcess): TtyConnector {
-                return this.createTtyConnector(process)
-            }
-        }
-
-        synchronized(this) {
-            val terminalOptions = ShellStartupOptions.Builder()
-                .workingDirectory(ModuleUtil.getModuleDirPath(module))
-                .shellCommand(initialShellCommand)
-                .build()
-            val process = terminalRunner.createProcess(terminalOptions)
-            val ttyConnector = terminalRunner.getTtyConnector(process)
-
-            currentProcess = process
-            currentConnector = ttyConnector
-            notifyProcessStarted(ttyConnector)
-        }
-    }
-
-    fun stopREPL() {
-        synchronized(this) {
-            currentProcess?.let {
-                it.destroy()
-                notifyProcessDestroyed()
-            }
-            currentProcess = null
-        }
-    }
-
-    val isRunning: Boolean
-        get() = currentProcess?.isAlive ?: false
-
-    fun addListener(listener: CommsEventListener) {
-        listeners.add(listener)
-
-        currentConnector?.let { listener.onProcessStarted(it) }
-    }
-
-    fun removeListener(listener: CommsEventListener) {
-        listeners.remove(listener)
-    }
-
-    private fun notifyProcessStarted(connector: TtyConnector) {
-        listeners.forEach { it.onProcessStarted(connector) }
-    }
-
-    private fun notifyProcessDestroyed() {
-        listeners.forEach { it.onProcessDestroyed() }
-    }
-
-    private fun notifyProcessCreationFailed(reason: String) {
-        listeners.forEach { it.onProcessCreationFailed(reason) }
-    }
 }
+
+val MICROPYTHON_REPL_CONTROL = Topic(MicroPythonReplControl::class.java)

--- a/src/main/kotlin/com/jetbrains/micropython/repl/MicroPythonReplManager.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/repl/MicroPythonReplManager.kt
@@ -2,9 +2,11 @@ package com.jetbrains.micropython.repl
 
 import com.intellij.openapi.components.Service
 import com.intellij.openapi.module.Module
+import com.intellij.openapi.module.ModuleUtil
 import com.jediterm.terminal.TtyConnector
 import com.jetbrains.micropython.settings.MicroPythonFacet
 import com.jetbrains.micropython.settings.microPythonFacet
+import com.jetbrains.python.sdk.baseDir
 import com.pty4j.PtyProcess
 import org.jetbrains.plugins.terminal.LocalTerminalDirectRunner
 import org.jetbrains.plugins.terminal.ShellStartupOptions
@@ -15,9 +17,7 @@ interface CommsEventListener {
     fun onProcessCreationFailed(reason: String)
 }
 
-@Service
-class MicroPythonReplManager(module: Module) {
-    private val currentModule: Module = module
+class MicroPythonReplManager(private val module: Module) {
     private var listeners: MutableList<CommsEventListener> = mutableListOf()
     private var currentProcess: Process? = null
     private var currentConnector: TtyConnector? = null
@@ -32,7 +32,7 @@ class MicroPythonReplManager(module: Module) {
             stopREPL()
         }
 
-        val facet = currentModule.microPythonFacet ?: return
+        val facet = module.microPythonFacet ?: return
         val devicePath = facet.getOrDetectDevicePathSynchronously()
 
         if (facet.pythonPath == null) {
@@ -51,7 +51,7 @@ class MicroPythonReplManager(module: Module) {
             devicePath
         )
 
-        val terminalRunner = object : LocalTerminalDirectRunner(currentModule.project) {
+        val terminalRunner = object : LocalTerminalDirectRunner(module.project) {
             override fun getInitialCommand(envs: MutableMap<String, String>): MutableList<String> {
                 return initialShellCommand
             }
@@ -63,7 +63,7 @@ class MicroPythonReplManager(module: Module) {
 
         synchronized(this) {
             val terminalOptions = ShellStartupOptions.Builder()
-                .workingDirectory(devicePath)
+                .workingDirectory(ModuleUtil.getModuleDirPath(module))
                 .shellCommand(initialShellCommand)
                 .build()
             val process = terminalRunner.createProcess(terminalOptions)

--- a/src/main/kotlin/com/jetbrains/micropython/repl/StopReplBeforeRunTask.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/repl/StopReplBeforeRunTask.kt
@@ -5,9 +5,8 @@ import com.intellij.execution.BeforeRunTaskProvider
 import com.intellij.execution.configurations.RunConfiguration
 import com.intellij.execution.runners.ExecutionEnvironment
 import com.intellij.openapi.actionSystem.DataContext
-import com.intellij.openapi.actionSystem.LangDataKeys
 import com.intellij.openapi.application.ApplicationManager
-import com.intellij.openapi.module.Module
+import com.intellij.openapi.components.service
 import com.intellij.openapi.util.Key
 import com.jetbrains.micropython.run.MicroPythonRunConfiguration
 import com.jetbrains.micropython.settings.MicroPythonFacetType
@@ -38,21 +37,14 @@ class StopReplBeforeRunTaskProvider : BeforeRunTaskProvider<StopReplBeforeRunTas
     }
 
     override fun executeTask(
-            context: DataContext,
-            configuration: RunConfiguration,
-            environment: ExecutionEnvironment,
-            task: StopReplBeforeRunTask
+        context: DataContext,
+        configuration: RunConfiguration,
+        environment: ExecutionEnvironment,
+        task: StopReplBeforeRunTask
     ): Boolean {
         if (configuration is MicroPythonRunConfiguration) {
-            ApplicationManager.getApplication().invokeLater {
-                configuration.module?.let {
-                    ApplicationManager.getApplication().runWriteAction {
-                        MicroPythonReplManager.getInstance(it).stopREPL()
-                    }
-                }
-            }
+            environment.project.service<MicroPythonReplManager>().stopRepl()
         }
-
         return true
     }
 

--- a/src/main/kotlin/com/jetbrains/micropython/repl/ToolWindowReplTab.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/repl/ToolWindowReplTab.kt
@@ -1,57 +1,62 @@
 package com.jetbrains.micropython.repl
 
 import com.intellij.icons.AllIcons
+import com.intellij.ide.ActivityTracker
 import com.intellij.openapi.Disposable
 import com.intellij.openapi.actionSystem.*
 import com.intellij.openapi.application.ApplicationManager
 import com.intellij.openapi.application.ModalityState
+import com.intellij.openapi.components.service
 import com.intellij.openapi.module.Module
-import com.intellij.openapi.project.DumbAware
+import com.intellij.openapi.project.DumbAwareAction
+import com.intellij.openapi.util.NlsContexts
+import com.intellij.util.application
 import com.jediterm.terminal.TtyConnector
 import com.jetbrains.micropython.settings.MicroPythonDevicesConfiguration
+import com.jetbrains.micropython.settings.MicroPythonFacet
+import com.jetbrains.micropython.settings.microPythonFacet
 import org.jetbrains.plugins.terminal.JBTerminalSystemSettingsProvider
+import org.jetbrains.plugins.terminal.LocalTerminalDirectRunner
+import org.jetbrains.plugins.terminal.ShellStartupOptions
 import org.jetbrains.plugins.terminal.ShellTerminalWidget
 import java.awt.BorderLayout
 import javax.swing.JPanel
 
-class ToolWindowReplTab(val module: Module, parent: Disposable) : CommsEventListener, Disposable {
+class ToolWindowReplTab(val module: Module, parent: Disposable) : MicroPythonReplControl {
     private val deviceConfiguration = MicroPythonDevicesConfiguration.getInstance(module.project)
-    private val deviceCommsManager = MicroPythonReplManager.getInstance(module)
     val terminalWidget: ShellTerminalWidget
 
     init {
         val mySettingsProvider = JBTerminalSystemSettingsProvider()
         terminalWidget = ShellTerminalWidget(module.project, mySettingsProvider, parent)
         terminalWidget.isEnabled = false
-        deviceCommsManager.addListener(this)
+        module.project.messageBus.connect(parent).subscribe(MICROPYTHON_REPL_CONTROL, this)
+
     }
 
     private fun connectWidgetTty(terminalWidget: ShellTerminalWidget, connector: TtyConnector) {
-        if (terminalWidget.isSessionRunning) {
-            terminalWidget.stop()
-        }
         terminalWidget.start(connector)
         val modalityState = ModalityState.stateForComponent(terminalWidget.component)
 
         ApplicationManager.getApplication().invokeLater(
-                {
-                    try {
-                        terminalWidget.component.revalidate()
-                        terminalWidget.notifyStarted()
-                    } catch (e: RuntimeException) {
-                        TODO("You can't cut back on error reporting! You will regret this!")
-                    }
-                },
-                modalityState
+            {
+                try {
+                    terminalWidget.component.revalidate()
+                    terminalWidget.notifyStarted()
+                } catch (e: RuntimeException) {
+                    TODO("You can't cut back on error reporting! You will regret this!")
+                }
+            },
+            modalityState
         )
     }
 
     fun createUI(): JPanel {
         val actionManager = ActionManager.getInstance()
         val toolbarActions = DefaultActionGroup().apply {
-            add(replStartAction())
-            add(replStopAction())
-            add(clearReplOnLaunch())
+            add(replStartAction)
+            add(replStopAction)
+            add(clearReplOnLaunch)
         }
         val actionToolbar = actionManager.createActionToolbar("MicroPythonREPL", toolbarActions, false)
         actionToolbar.targetComponent = terminalWidget.component
@@ -64,39 +69,36 @@ class ToolWindowReplTab(val module: Module, parent: Disposable) : CommsEventList
         }
     }
 
-    private fun replStopAction() = object : AnAction(
-            "Stop", "Stop REPL session", AllIcons.Actions.Suspend
-    ), DumbAware {
+    private val replStopAction = object : DumbAwareAction(
+        "Stop", "Stop REPL session", AllIcons.Actions.Suspend
+    ) {
         override fun update(e: AnActionEvent) {
-            e.presentation.isEnabled = deviceCommsManager.isRunning
+            e.presentation.isEnabled = terminalWidget.isSessionRunning
         }
 
         override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.EDT
 
         override fun actionPerformed(e: AnActionEvent) {
-            terminalWidget.stop()
-            deviceCommsManager.stopREPL()
+            stopRepl()
         }
     }
 
-    private fun replStartAction() =
-        object : AnAction("Restart", "Restart REPL session", AllIcons.Actions.Restart), DumbAware {
+    private val replStartAction =
+        object : DumbAwareAction("Restart", "Restart REPL session", AllIcons.Actions.Restart) {
             override fun update(e: AnActionEvent) {
                 e.presentation.isEnabled = true
             }
 
             override fun getActionUpdateThread(): ActionUpdateThread = ActionUpdateThread.EDT
 
-            override fun actionPerformed(e: AnActionEvent) {
-                if (terminalWidget.isSessionRunning) {
-                    deviceCommsManager.stopREPL()
-                }
-                deviceCommsManager.startREPL()
-            }
+            override fun actionPerformed(e: AnActionEvent) =
+                module.project.service<MicroPythonReplManager>().startOrRestartRepl()
         }
 
-    private fun clearReplOnLaunch() = object : ToggleAction("Clear Window On Start",
-            "Clear REPL window on every start", AllIcons.Actions.GC) {
+    private val clearReplOnLaunch = object : ToggleAction(
+        "Clear Window On Start",
+        "Clear REPL window on every start", AllIcons.Actions.GC
+    ) {
         override fun isSelected(e: AnActionEvent): Boolean {
             return deviceConfiguration.clearReplOnLaunch
         }
@@ -110,31 +112,76 @@ class ToolWindowReplTab(val module: Module, parent: Disposable) : CommsEventList
         }
     }
 
-    override fun dispose() {
-        deviceCommsManager.removeListener(this)
-    }
-
-    override fun onProcessStarted(ttyConnector: TtyConnector) {
-        if (deviceConfiguration.clearReplOnLaunch) {
-            terminalWidget.terminalTextBuffer.clearHistory()
-            terminalWidget.terminal.reset()
-        } else {
-            terminalWidget.terminal.nextLine()
-        }
-        connectWidgetTty(terminalWidget, ttyConnector)
-        terminalWidget.isEnabled = true
-    }
-
-    override fun onProcessDestroyed() {
-        terminalWidget.stop()
-
-        terminalWidget.terminal.nextLine()
-        terminalWidget.terminal.writeCharacters("=== SESSION HAS BEEN INTERRUPTED ===")
-        terminalWidget.terminal.nextLine()
-    }
-
-    override fun onProcessCreationFailed(reason: String) {
+    private fun onProcessCreationFailed(@NlsContexts.SystemNotificationText reason: String) {
         terminalWidget.terminal.nextLine()
         terminalWidget.terminal.writeCharacters(reason)
+    }
+
+    override fun stopRepl() {
+        application.executeOnPooledThread {
+            synchronized(this) {
+                terminalWidget.processTtyConnector?.process?.destroy()
+            }
+        }
+    }
+
+    override fun startOrRestartRepl() {
+        if (terminalWidget.isSessionRunning) {
+            stopRepl()
+        }
+
+        val facet = module.microPythonFacet ?: return
+        val devicePath = facet.getOrDetectDevicePathSynchronously()
+
+        if (facet.pythonPath == null) {
+            onProcessCreationFailed("Valid Python interpreter is needed to start REPL!")
+            return
+        }
+
+        if (devicePath == null) {
+            onProcessCreationFailed("Device path is not specified, please check settings.")
+            return
+        }
+
+        val initialShellCommand = mutableListOf(
+            facet.pythonPath!!,
+            "${MicroPythonFacet.scriptsPath}/microrepl.py",
+            devicePath
+        )
+
+        val terminalRunner = LocalTerminalDirectRunner(module.project)
+
+        synchronized(this) {
+            val terminalOptions = terminalRunner.configureStartupOptions(
+                ShellStartupOptions.Builder()
+                    .shellCommand(initialShellCommand)
+                    .build()
+            )
+
+            val process = terminalRunner.createProcess(terminalOptions)
+            val ttyConnector = terminalRunner.createTtyConnector(process)
+            process.onExit().whenComplete { _, _ ->
+                application.invokeLater(
+                    {
+                        ActivityTracker.getInstance().inc()
+                        with(terminalWidget.terminal) {
+                            nextLine()
+                            writeCharacters("=== SESSION HAS BEEN INTERRUPTED ===")
+                            nextLine()
+                        }
+                    },
+                    { module.isDisposed }
+                )
+            }
+
+            if (deviceConfiguration.clearReplOnLaunch) {
+                terminalWidget.terminalTextBuffer.clearHistory()
+                terminalWidget.terminal.reset()
+            } else {
+                terminalWidget.terminal.nextLine()
+            }
+            connectWidgetTty(terminalWidget, ttyConnector)
+            terminalWidget.isEnabled = true
+        }
     }
 }

--- a/src/main/kotlin/com/jetbrains/micropython/repl/ToolWindowReplTab.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/repl/ToolWindowReplTab.kt
@@ -20,6 +20,7 @@ import org.jetbrains.plugins.terminal.LocalTerminalDirectRunner
 import org.jetbrains.plugins.terminal.ShellStartupOptions
 import org.jetbrains.plugins.terminal.ShellTerminalWidget
 import java.awt.BorderLayout
+import java.util.concurrent.TimeUnit
 import javax.swing.JPanel
 
 class ToolWindowReplTab(val module: Module, parent: Disposable) : MicroPythonReplControl {
@@ -117,7 +118,21 @@ class ToolWindowReplTab(val module: Module, parent: Disposable) : MicroPythonRep
         terminalWidget.terminal.writeCharacters(reason)
     }
 
+    private fun interruptBanner() {
+        application.invokeLater(
+            {
+                with(terminalWidget.terminal) {
+                    nextLine()
+                    writeCharacters("=== SESSION HAS BEEN INTERRUPTED ===")
+                    nextLine()
+                }
+            },
+            { module.isDisposed }
+        )
+
+    }
     override fun stopRepl() {
+        interruptBanner()
         application.executeOnPooledThread {
             synchronized(this) {
                 terminalWidget.processTtyConnector?.process?.destroy()
@@ -126,10 +141,21 @@ class ToolWindowReplTab(val module: Module, parent: Disposable) : MicroPythonRep
     }
 
     override fun startOrRestartRepl() {
-        if (terminalWidget.isSessionRunning) {
-            stopRepl()
+        interruptBanner()
+        application.executeOnPooledThread {
+            synchronized(this) {
+                terminalWidget.processTtyConnector?.process?.apply {
+                    if (isAlive) destroy()
+                    waitFor(10, TimeUnit.SECONDS)
+                }
+            }
+            application.invokeLater(
+                { startRepl() },
+                { module.project.isDisposed })
         }
+    }
 
+    private fun startRepl() {
         val facet = module.microPythonFacet ?: return
         val devicePath = facet.getOrDetectDevicePathSynchronously()
 
@@ -160,20 +186,7 @@ class ToolWindowReplTab(val module: Module, parent: Disposable) : MicroPythonRep
 
             val process = terminalRunner.createProcess(terminalOptions)
             val ttyConnector = terminalRunner.createTtyConnector(process)
-            process.onExit().whenComplete { _, _ ->
-                application.invokeLater(
-                    {
-                        ActivityTracker.getInstance().inc()
-                        with(terminalWidget.terminal) {
-                            nextLine()
-                            writeCharacters("=== SESSION HAS BEEN INTERRUPTED ===")
-                            nextLine()
-                        }
-                    },
-                    { module.isDisposed }
-                )
-            }
-
+            process.onExit().whenComplete { _, _ -> ActivityTracker.getInstance().inc() }
             if (deviceConfiguration.clearReplOnLaunch) {
                 terminalWidget.terminalTextBuffer.clearHistory()
                 terminalWidget.terminal.reset()

--- a/src/main/kotlin/com/jetbrains/micropython/run/MicroPythonRunConfiguration.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/run/MicroPythonRunConfiguration.kt
@@ -30,6 +30,7 @@ import com.intellij.execution.runners.ProgramRunner
 import com.intellij.facet.ui.ValidationResult
 import com.intellij.openapi.actionSystem.LangDataKeys
 import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.components.service
 import com.intellij.openapi.module.Module
 import com.intellij.openapi.module.ModuleUtil
 import com.intellij.openapi.options.ShowSettingsUtil
@@ -39,7 +40,6 @@ import com.intellij.openapi.util.Key
 import com.intellij.openapi.util.text.StringUtil
 import com.intellij.openapi.vfs.StandardFileSystems
 import com.intellij.openapi.wm.ToolWindowManager
-import com.intellij.serviceContainer.ComponentManagerImpl
 import com.intellij.util.PathUtil
 import com.intellij.util.PlatformUtils
 import com.jetbrains.micropython.repl.MicroPythonReplManager
@@ -87,7 +87,7 @@ class MicroPythonRunConfiguration(project: Project, factory: ConfigurationFactor
     if (runReplOnSuccess && state != null) {
       return RunStateWrapper(state) {
         ApplicationManager.getApplication().invokeLater {
-          MicroPythonReplManager.getInstance(currentModule).startREPL()
+          project.service<MicroPythonReplManager>().startOrRestartRepl()
           ToolWindowManager.getInstance(project).getToolWindow("MicroPython")?.show()
         }
       }

--- a/src/main/kotlin/com/jetbrains/micropython/ui/MicroPythonToolWindowFactory.kt
+++ b/src/main/kotlin/com/jetbrains/micropython/ui/MicroPythonToolWindowFactory.kt
@@ -1,5 +1,6 @@
 package com.jetbrains.micropython.ui
 
+import com.intellij.openapi.components.service
 import com.intellij.openapi.project.DumbAware
 import com.intellij.openapi.project.Project
 import com.intellij.openapi.wm.ToolWindow
@@ -16,12 +17,9 @@ class MicroPythonToolWindowFactory : ToolWindowFactory, DumbAware {
 
         project.firstMicroPythonFacet?.let {
             terminalContent.component = ToolWindowReplTab(it.module, terminalContent).createUI()
-        }
-
-        toolWindow.contentManager.addContent(terminalContent)
-
-        project.firstMicroPythonFacet?.let {
-            MicroPythonReplManager.getInstance(it.module).startREPL()
+            toolWindow.contentManager.addContent(terminalContent)
+            project.service<MicroPythonReplManager>().startOrRestartRepl()
         }
     }
+
 }

--- a/src/main/resources/META-INF/plugin.xml
+++ b/src/main/resources/META-INF/plugin.xml
@@ -42,7 +42,6 @@
                          displayName="MicroPython"
                          provider="com.jetbrains.micropython.settings.MicroPythonConfigurableProvider"/>
     <projectService serviceImplementation="com.jetbrains.micropython.settings.MicroPythonDevicesConfiguration"/>
-    <moduleService serviceImplementation="com.jetbrains.micropython.repl.MicroPythonReplManager"/>
     <stepsBeforeRunProvider implementation="com.jetbrains.micropython.repl.StopReplBeforeRunTaskProvider"/>
     <localInspection language="Python"
                      displayName="MicroPython requirements inspection"

